software-mansion/radon-ide
# Add option to capture replays

This PR adds replays support.

Replays allows the user to capture a recent moment into a video and preview it instantly within the IDE panel.

The current way this feature has been designed is as follows:
 - in device setting you can enable/disable replays functionality. Disabling hides the replays button from the UI
 - we're adding new "replay" button that's in the main IDE panel and can be use to capture the replay and moves the user into the preview mode
 - in preview mode we're displaying a full screen overlay over the IDE panel with video player
 - the overlay has a seekbar, arrows to move forward/backward by one frame, save button, and the video length selector
 - simulator-server uses a fixed size memory buffer for replays, depending on the frame rate (it is variable) it may store a different lengths of the video depending on the content. When replay is captured we show the past 5 seconds by default and allow user to increase that length using length selector as long as the captured video is of a sufficient duration.


Here's what we are adding as part of this PR:
1) New UI element to PreviewView that displays the preview button
2) New UI screen that's responsible for rendering replay preview with all the controls
3) We're extending the webview<>extension API with methods for capturing video replays and for saving video files
4) We're extensing device settings with the new replay setting. Then we use this setting to start/stop recording depending on device session state. We want the recording to start after every device reload, such that we're not recording reload screens etc.

## Test plan

1. Enable replays in device options -> see it changes the appearance of replay button in top-right
2. Tap on replay after a while, the video should appear in an overlay
3. For longer videos, you should be able to change between lengths
4. Updating length once video finished playback should make it play from a certain moment
5. Updating length to shorter when player is at the start of playback, should move to the start of new length
6. Save button should store the full length video. When one is stored with the same name, you'll get a confirmation dialog
